### PR TITLE
IGNITE-13754 .NET: Fix LINQ provider for queries with JOIN and GROUP BY combined

### PR DESCRIPTION
**Pull request by ptupitsyn**
Fix projection traversal in `AliasDictionary.GetQuerySource` and `CacheQueryExpressionVisitor.VisitMember`: use `memberHint` to understand where the given member of the projected type comes from to emit correct table and field name.

### CODE REVIEW

File: modules/platforms/dotnet/Apache.Ignite.Linq/Impl/CacheQueryExpressionVisitor.cs [hunk 1]
```diff
@@ -676,5 +683,20 @@ private void VisitArguments(IEnumerable<Expression> arguments)
...
             }
         }
+
+        /// <summary>
+        /// Determines whether specified type is an anonymous type.
+        /// <para />
+        /// There is no 100% way to identify an anonymous type, since it is up to the compiler implementation,
+        /// so we rely on some well-known anonymous type properties.
+        /// </summary>
+        private static bool IsAnonymousType(Type type)
```
Comment (gurustron) [inline modules/platforms/dotnet/Apache.Ignite.Linq/Impl/CacheQueryExpressionVisitor.cs:693]: Maybe add check for `CompilerGeneratedAttribute` ?

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Cache/Query/Linq/CacheLinqTest.GroupBy.cs [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+            // Ordering and count
+            var res1 =
+                from p in persons
+                orderby p.Value.Name
+                group p by p.Value.OrganizationId
+                into gs
+                orderby gs.Key
+                where gs.Count() > 10
+                select new { Count = gs.Count(), OrgId = gs.Key, AvgAge = gs.Average(x => x.Value.Age) };
```
Comment (gurustron) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Cache/Query/Linq/CacheLinqTest.GroupBy.cs:104]: This throws if we change it to:

```
        public class MyClass
        {
            public int Count { get; set; }
            public int OrgId { get; set; }
            public double AvgAge { get; set; }
        }

        .....
       select new MyClass { Count = ....

```

Should we support such scenario also?

Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Cache/Query/Linq/CacheLinqTest.GroupBy.cs:104]: Good point, we should not actually limit our projection support to anonymous types.
* Anonymous type check removed
* Custom class projection test added
* Member init expression support added

Comment (gurustron) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Cache/Query/Linq/CacheLinqTest.GroupBy.cs:104]: @ptupitsyn looks good to me. 
```

**Pull request merged**